contracts/psp22/lib.rs

```diff
@@ -0,0 +1,153 @@
+            }
+
+            self.balances.insert(from, &(from_balance - value));
+            let to_balance = self.balance_of(*to);
+            self.balances.insert(to, &(to_balance + value));
+
+            Self::emit_event(
+                self.env(),
+                Event::Transfer(Transfer {
+                    from: Some(*from),
+                    to: Some(*to),
+                    value,
+                }),
+            );
+
+            Ok(())
+        }
+
+        fn emit_event<EE>(emitter: EE, event: Event)
+        where
+            EE: EmitEvent<Self>,
+        {
+            emitter.emit_event(event);
+        }
+    }
+
+    impl PSP22 for Token {
+        /// Returns the total token supply.
+        #[ink(message)]
+        fn total_supply(&self) -> Balance {
+            self.total_supply
+        }
+
+        /// Returns the account balance for the specified `owner`.
+        #[ink(message)]
+        fn balance_of(&self, owner: AccountId) -> Balance {
+            self.balances.get(owner).unwrap_or_default()
+        }
+
+        /// Returns the amount which `spender` is allowed to withdraw on behalf of the `owner` account.
+        #[ink(message)]
+        fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+            self.allowances.get((owner, spender)).unwrap_or_default()
+        }
+
+        /// Allows `spender` to withdraw from the caller's account multiple times, up to the `value` amount.
+        #[ink(message)]
+        fn approve(&mut self, spender: AccountId, value: Balance) -> Result<(), PSP22Error> {
+            let owner = self.env().caller();
+            self.allowances.insert((&owner, &spender), &value);
+
+            Self::emit_event(
+                self.env(),
+                Event::Approval(Approval {
+                    owner,
+                    spender,
+                    value,
+                }),
+            );
+
+            Ok(())
+        }
+
+        /// Transfers `value` amount of tokens from the caller's account to account `to`.
+        #[ink(message)]
+        fn transfer(&mut self, to: AccountId, value: Balance) -> Result<(), PSP22Error> {
+            let from = self.env().caller();
+            self.transfer_from_to(&from, &to, value)
+        }
+
+        /// Transfers `value` amount of tokens on the behalf of `from` to the account `to`.
+        /// Caller need to be pre-approved
+        #[ink(message)]
+        fn transfer_from(
+            &mut self,
+            from: AccountId,
+            to: AccountId,
+            value: Balance,
+        ) -> Result<(), PSP22Error> {
+            let caller = self.env().caller();
+            let allowance = self.allowance(from, caller);
+
+            if allowance < value {
+                return Err(PSP22Error::InsufficientAllowance);
+            }
+
+            self.transfer_from_to(&from, &to, value)?;
+
+            self.allowances
+                .insert((&from, &caller), &(allowance - value));
```

Comment:
we could add check for `allowance == value` and if so, remove entry

Reply:
yeah nice little storage optimization (very little) - maybe leave it for students to spot?